core/src/main/scala/kafka/utils/NetworkClientBlockingOps.scala

```diff
@@ -0,0 +1,102 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.utils
+
+import java.io.IOException
+import org.apache.kafka.clients.{ConnectionState, ClientRequest, ClientResponse, NetworkClient}
+import org.apache.kafka.common.Node
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+
+import org.apache.kafka.common.utils.{Time => JTime}
+
+object NetworkClientBlockingOps {
+  implicit def networkClientBlockingOps(client: NetworkClient): NetworkClientBlockingOps =
+    new NetworkClientBlockingOps(client)
+}
+
+class NetworkClientBlockingOps(val client: NetworkClient) extends AnyVal {
+
+  def blockingReady(node: Node, timeout: Long, retryBackoff: Option[Long])(implicit time: JTime): Boolean = {
+    client.ready(node, time.milliseconds()) || pollUntil(timeout) { (_, now) =>
+      if (client.isReady(node, now))
+        true
+      else if (client.connectionFailed(node))
+        throw new IOException(s"Connection to $node failed")
+      else {
+        retryBackoff.foreach(Thread.sleep)
+        false
+      }
+    }
+  }
+
+  def disconnectAndPoll(nodeId: String, pollTimeout: Long)(implicit time: JTime): Boolean = {
+    client.disconnect(nodeId)
+    client.poll(pollTimeout, time.milliseconds())
+    client.connectionState(nodeId) == ConnectionState.DISCONNECTED
+  }
+
+  def blockingSendAndReceive(request: ClientRequest, timeout: Long)(implicit time: JTime): Option[ClientResponse] = {
+    client.send(request)
```

Comment:
Yes, I understand. So, if the connection is not ready, an `IllegalStateException` will be thrown which will cause a disconnect (no-op) followed by another connect and then we will retry again after a 300ms sleep since `isSendSuccessful === false`. This behaviour is very similar to what would happen with `BlockingChannel` where a ClosedChannelException would be thrown if `send` was called and the channel was not connected due to a failure.

I can add a `blockingReady` call before the `send` if you think that is better though.